opm/simulators/linalg/DILU.hpp

```diff
@@ -0,0 +1,183 @@
+
+/*
+  Copyright 2022-2023 SINTEF AS
+  This file is part of the Open Porous Media project (OPM).
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_DILU_HEADER_INCLUDED
+#define OPM_DILU_HEADER_INCLUDED
+
+#include <config.h>
+#include <opm/common/ErrorMacros.hpp>
+#include <opm/common/TimingMacros.hpp>
+#include <opm/simulators/linalg/PreconditionerWithUpdate.hpp>
+
+#include <dune/common/fmatrix.hh>
+#include <dune/common/version.hh>
+#include <dune/common/unused.hh>
+#include <dune/istl/bcrsmatrix.hh>
+
+
+namespace Dune
+{
+
+/*! \brief The sequential DILU preconditioner.
+
+   \tparam M The matrix type to operate on
+   \tparam X Type of the update
+   \tparam Y Type of the defect
+*/
+template <class M, class X, class Y>
+class SeqDilu : public PreconditionerWithUpdate<X, Y>
+{
+    public:
+        //! \brief The matrix type the preconditioner is for.
+        using matrix_type = M;
+        //! \brief The domain type of the preconditioner.
+        using domain_type = X;
+        //! \brief The range type of the preconditioner.
+        using range_type = Y;
+        //! \brief The field type of the preconditioner.
+        using field_type = typename X::field_type;
+        //! \brief scalar type underlying the field_type
+
+    /*! \brief Constructor.
+       Constructor gets all parameters to operate the prec.
+       \param A The matrix to operate on.
+    */
+    SeqDilu(const M& A)
+        : A_(A)
+    {
+        Dinv_.resize(A_.N());
+        // the Dinv matrix must be initialised
+        update();
+    }
+
+    /*!
+       \brief Update the preconditioner.
+       \copydoc Preconditioner::update()
+    */
+    virtual void update() override
+    {
+        OPM_TIMEBLOCK(update);
+        
+        auto endi = A_.end();
+        for ( auto row_i = A_.begin(); row_i != endi; ++row_i) {
+            Dinv_[row_i.index()] = A_[row_i.index()][row_i.index()];
+        }
+        
+        for ( auto row_i = A_.begin(); row_i != endi; ++row_i)
+        {
+           auto Dinv_temp = Dinv_[row_i.index()];
+            for (auto a_ij = row_i->begin(); a_ij.index() < row_i.index(); ++a_ij) {
+                auto a_ji = A_[a_ij.index()].find(row_i.index());
```

Comment:
Maybe add a local variable `const auto col_j = a_ij.index()` or similar since that is used a lot, I think it would be easier to read.

Also, you could use `const auto a_ji = ...`.

Reply:
Good suggestion, this will increase the readability. I changed this now.